waiter/src/waiter/metrics_consumer.clj

```diff
@@ -0,0 +1,222 @@
+          query-chan (async/chan)
+          default-cluster (token/get-default-cluster token-cluster-calculator)
+          filter-events-fn
+          (fn filter-events-fn
+            [{:strs [token]}]
+            (let [token-parameters (sd/token->token-parameters kv-store token :error-on-missing false)
+                  token-cluster (get token-parameters "cluster")]
+              (and (some? token-parameters)
+                   (= token-cluster default-cluster))))
+          map-events-fn
+          (fn map-events-fn
+            [{:strs [lastRequestTime token] :as event}]
+            (timers/start-stop-time!
+              (metrics/waiter-timer "core" "metrics-consumer" "map-events-fn")
+              (when-let [{:strs [run-as-user] :as service-description-template}
+                         (sd/token->service-parameter-template kv-store token :error-on-missing false)]
+                (when (and run-as-user
+                           (not (sd/run-as-requester? service-description-template))
+                           (not (sd/requires-parameters? service-description-template)))
+                  (try
+                    (let [{:keys [descriptor latest-descriptor]} (retrieve-descriptor-fn run-as-user token)]
+                      (assoc event "lastRequestTime" (f/parse lastRequestTime)
+                                   "descriptor" descriptor
+                                   "latest-descriptor" latest-descriptor))
+                    (catch Exception e
+                      (log/error e "unable to retrieve descriptor" {:run-as-user run-as-user
+                                                                    :token token})))))))
+          metrics-service-watches
+          (doall
+            (map
+              (fn [metrics-service-url]
+                (make-metrics-watch-request-fn
+                  router-id metrics-service-url token-metric-chan filter-events-fn map-events-fn
+                  streaming-api-request clock connection-timeout-ms idle-timeout-ms retry-delay-ms))
+              metrics-service-urls))
+          query-state-fn
+          (fn query-metrics-consumer-maintainer-fn
+            [include-flags]
+            (let [{:keys [last-update-time]} @state-atom]
+              (cond-> {:last-update-time last-update-time
+                       :supported-include-params ["buffer-state" "watches-state"]}
+                (contains? include-flags "buffer-state")
+                (assoc :buffer-state {:envoy-cluster-event-count (.count token-metric-chan-buffer)})
+                (contains? include-flags "watches-state")
+                (assoc :watches-state (reduce
+                                        (fn [watches-state {:keys [metrics-service-url query-state-fn]}]
+                                          (assoc watches-state metrics-service-url (query-state-fn include-flags)))
+                                        {}
+                                        metrics-service-watches)))))
```

Comment:
If `metrics-service-watches` is a map keyed by the `metrics-service-url`, we can avoid the `reduce` and replace with a `pc/map-vals`.